core/src/main/java/org/apache/calcite/sql/type/SqlTypeUtil.java

```diff
@@ -123,21 +126,13 @@ public static boolean isCharTypeComparable(
       SqlCallBinding binding,
       List<SqlNode> operands,
       boolean throwOnFailure) {
-    final SqlValidator validator = binding.getValidator();
-    final SqlValidatorScope scope = binding.getScope();
-    assert operands != null;
-    assert operands.size() >= 2;
+    requireNonNull(operands, "operands");
+    assert operands.size() >= 2
```

Comment:
I think we should avoid the use of `assert` due to the assertion is usually disabled in a JVM.

Reply:
The code was there long ago (I guess before `Objects.requireNonNull` existed), so it was using asserts.
However,
1) Significant rework of `assert` is out of the scope of this PR
2) `assert operands.size() >= 2` does not seem to have a corresponding Java API.

Reply:
> assert operands.size() >= 2 does not seem to have a corresponding Java API.
I agree that we should keep the scope of this PR minimal and clear, and maybe rework the `assert` in another PR,
but regrading to the above comments, I think we can use the guava's `Preconditions.checkArgument` to do this check.

Reply:
`Preconditions.checkArgument` allocates garbage even in case the error is not thrown

Reply:
hi @vlsi , what do you mean by `allocates garbage`?

Reply:
I mean primitive arguments boxing, and vararg array allocation.

This is missing in Guava 19: https://github.com/google/guava/commit/892e323fca32945cdfb25395ca6e346dd0fffa5b#diff-fe7358934fa6eba23c2791eb40cec030